gcsio/src/main/java/com/google/cloud/hadoop/gcsio/BatchExecutor.java

```diff
@@ -0,0 +1,144 @@
+
+import com.google.cloud.hadoop.util.GoogleCloudStorageEventBus;
+import com.google.common.flogger.GoogleLogger;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.io.IOException;
+import java.util.Queue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * BatchExecutor provides a means to manually batch requests using a thread pool. Execution is
+ * performed by the underlying {@link #requestsExecutor} ExecutorService.
+ *
+ * <p>Expected usage is to create a new BatchExecutor instance per client operation that represents
+ * logical grouping of requests(delete, copy, get).
+ *
+ * <p>Instance of this class can not be used again after {@link #shutdown()} method has been called.
+ */
+class BatchExecutor {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private final ExecutorService requestsExecutor;
+
+  private final Queue<Future<Void>> responseFutures = new ConcurrentLinkedQueue<>();
+
+  public BatchExecutor(int numThreads) {
+    this.requestsExecutor =
+        numThreads == 0 ? newDirectExecutorService() : newRequestExecutor(numThreads);
+  }
+
+  private static ExecutorService newRequestExecutor(int numThreads) {
+    ThreadPoolExecutor executor =
+        new ThreadPoolExecutor(
+            /* corePoolSize= */ numThreads,
+            /* maximumPoolSize= */ numThreads,
+            /* keepAliveTime= */ 10L,
+            TimeUnit.SECONDS,
+            new LinkedBlockingQueue<>(numThreads * 20),
+            new ThreadFactoryBuilder()
+                .setNameFormat("gcs-manual-batching-pool-%d")
+                .setDaemon(true)
+                .build());
+    executor.allowCoreThreadTimeOut(true);
+    executor.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
+    return executor;
+  }
+
+  /** Adds a task to the execution queue. */
+  public <T> void queue(Callable<T> task, FutureCallback<T> callback) {
+    checkState(
+        !requestsExecutor.isShutdown() && !requestsExecutor.isTerminated(),
+        "requestExecutor should not be terminated to queue request");
+
+    responseFutures.add(
+        requestsExecutor.submit(
+            () -> {
+              execute(task, callback);
+              return null;
+            }));
+  }
+
+  private static <T> void execute(Callable<T> task, FutureCallback<T> callback) throws Exception {
+    checkArgument(callback != null, "FutureCallBack cannot be null : %s", callback);
+    try {
+      T result = task.call();
+      callback.onSuccess(result);
+    } catch (Throwable throwable) {
+      callback.onFailure(throwable);
```

Comment:
It does not look like this can throw any exception

Reply:
Correct